allennlp/modules/elmo.py

```diff
@@ -499,14 +532,31 @@ def forward(self,  # pylint: disable=arguments-differ
         Note that the output tensors all include additional special begin and end of sequence
         markers.
         """
-        token_embedding = self._token_embedder(inputs)
-        type_representation = token_embedding['token_embedding']
-        mask = token_embedding['mask']
+        if self._word_embedding is not None:
+            mask_without_bos_eos = (inputs > 0).long()
+            # The character cnn part is cached - just look it up.
+            embedded_inputs = self._word_embedding(inputs) # type: ignore
+            # shape (batch_size, timesteps + 2, embedding_dim)
+            type_representation, mask = add_sentence_boundary_token_ids(
+                    embedded_inputs,
+                    mask_without_bos_eos,
+                    self._bos_embedding,
+                    self._eos_embedding
+            )
+        else:
+            token_embedding = self._token_embedder(inputs)
+            mask = token_embedding['mask']
+            type_representation = token_embedding['token_embedding']
         lstm_outputs = self._elmo_lstm(type_representation, mask)
 
         # Prepare the output.  The first layer is duplicated.
+        # Because of minor differences in how masking is applied depending
+        # on whether the char cnn layers are cached, we'll be defensive and
+        # multiply by the mask here. It's not strictly necessary, as the
+        # mask passed on is correct, but the values in the padded areas
+        # of the char cnn representations can change.
```

Comment:
Wow, that sounds like it was a pain.  I assume you found this because you were testing for exactly reproducing the outputs between the two branches?

Reply:
Yes